lua/nvim-tree/renderer/decorator/git.lua

```diff
@@ -0,0 +1,205 @@
+local notify = require "nvim-tree.notify"
+local explorer_node = require "nvim-tree.explorer.node"
+
+local HL_POSITION = require("nvim-tree.enum").HL_POSITION
+local ICON_PLACEMENT = require("nvim-tree.enum").ICON_PLACEMENT
+
+local Decorator = require "nvim-tree.renderer.decorator"
+
+--- @class DecoratorGit: Decorator
+--- @field enabled boolean
+--- @field file_hl string[]
+--- @field folder_hl string[]
+--- @field git_icons table
+local DecoratorGit = Decorator:new()
+
+local function build_icons_table(i)
+  local icons = {
+    staged = { str = i.staged, hl = { "NvimTreeGitStagedIcon" }, ord = 1 },
+    unstaged = { str = i.unstaged, hl = { "NvimTreeGitDirtyIcon" }, ord = 2 },
+    renamed = { str = i.renamed, hl = { "NvimTreeGitRenamedIcon" }, ord = 3 },
+    deleted = { str = i.deleted, hl = { "NvimTreeGitDeletedIcon" }, ord = 4 },
+    unmerged = { str = i.unmerged, hl = { "NvimTreeGitMergeIcon" }, ord = 5 },
+    untracked = { str = i.untracked, hl = { "NvimTreeGitNewIcon" }, ord = 6 },
+    ignored = { str = i.ignored, hl = { "NvimTreeGitIgnoredIcon" }, ord = 7 },
+  }
+  return {
+    ["M "] = { icons.staged },
+    [" M"] = { icons.unstaged },
+    ["C "] = { icons.staged },
+    [" C"] = { icons.unstaged },
+    ["CM"] = { icons.unstaged },
+    [" T"] = { icons.unstaged },
+    ["T "] = { icons.staged },
+    ["TM"] = { icons.staged, icons.unstaged },
+    ["MM"] = { icons.staged, icons.unstaged },
+    ["MD"] = { icons.staged },
+    ["A "] = { icons.staged },
+    ["AD"] = { icons.staged },
+    [" A"] = { icons.untracked },
+    -- not sure about this one
+    ["AA"] = { icons.unmerged, icons.untracked },
+    ["AU"] = { icons.unmerged, icons.untracked },
+    ["AM"] = { icons.staged, icons.unstaged },
+    ["??"] = { icons.untracked },
+    ["R "] = { icons.renamed },
+    [" R"] = { icons.renamed },
+    ["RM"] = { icons.unstaged, icons.renamed },
+    ["UU"] = { icons.unmerged },
+    ["UD"] = { icons.unmerged },
+    ["UA"] = { icons.unmerged },
+    [" D"] = { icons.deleted },
+    ["D "] = { icons.deleted },
+    ["DA"] = { icons.unstaged },
+    ["RD"] = { icons.deleted },
+    ["DD"] = { icons.deleted },
+    ["DU"] = { icons.deleted, icons.unmerged },
+    ["!!"] = { icons.ignored },
+    dirty = { icons.unstaged },
+  }
+end
+
+local function build_hl_table()
+  local file = {
+    ["M "] = "NvimTreeGitFileStagedHL",
+    ["C "] = "NvimTreeGitFileStagedHL",
+    ["AA"] = "NvimTreeGitFileStagedHL",
+    ["AD"] = "NvimTreeGitFileStagedHL",
+    ["MD"] = "NvimTreeGitFileStagedHL",
+    ["T "] = "NvimTreeGitFileStagedHL",
+    ["TT"] = "NvimTreeGitFileStagedHL",
+    [" M"] = "NvimTreeGitFileDirtyHL",
+    ["CM"] = "NvimTreeGitFileDirtyHL",
+    [" C"] = "NvimTreeGitFileDirtyHL",
+    [" T"] = "NvimTreeGitFileDirtyHL",
+    ["MM"] = "NvimTreeGitFileDirtyHL",
+    ["AM"] = "NvimTreeGitFileDirtyHL",
+    dirty = "NvimTreeGitFileDirtyHL",
+    ["A "] = "NvimTreeGitFileNewHL",
```

Comment:
I found a bug here for that I made a pr (2534) , I see you moved the logic from components to decorator, if my pr don't pass, please can you fix this one in your pr? ["A "] = { icons.staged }, vs ["A "] = "NvimTreeFileStaged", can be reproduced: create a new file in your git repository and after that stage it (git add filename), I think the correct state is staged icons are good but highlight not.

Reply:
Many thanks @tomcri !

I'll make sure that fix is not lost...